File: prolog/ts.pl
% program(Name, InState, InSymbol, OutState, OutSymbol, Direction).
program(plus1, q0, 1, q0, 1, right).
program(plus1, q0, b, final, 1, stay).

get_last_or_b([], b).
get_last_or_b([H | []], H).
get_last_or_b([_|T], E) :-
    T \= [],
    get_last_or_b(T, E).

get_first_or_b([], b).   
get_first_or_b([H | _], H).
    
remove_last([], []).
remove_last([_], []).
remove_last([H1 | T1], L2) :-
    remove_last(T1, Rez),
    L2 = [H1 | Rez].

remove_first([], []).
remove_first([_ | T], T).

reverse_list([], []).
reverse_list([H | T], L2) :-
    reverse_list(T, Rez),
    append(Rez, [H], L2).
    
% action(Direction, InL-InR, OutL-OutR)
action(stay, InL-InR, InL-InR).

action(right, InL-InR, OutL-OutR) :-
    get_first_or_b(InR, ElPodGlavo),
    remove_first(InR, OutR),
    OutL = [ElPodGlavo | InL].

action(left, InL-InR, OutL-OutR) :-
    get_first_or_b(InL, ZadnjiInL),
    OutR = [ZadnjiInL | InR],
    remove_first(InL, OutL).
    
% head_rest(R, Head, Rest)
head_rest([], b, []).
head_rest([RHead | RTail], RHead, RTail).
    
% step(Name, InL-InR, InState, OutL-OutR, OutState)
step(Name, InL-InR, InState, OutL-OutR, OutState) :-
    head_rest(InR, InSymbol, InRTail),
    program(Name, InState, InSymbol, OutState, OutSymbol, Direction),
    action(Direction, InL-[OutSymbol | InRTail], OutL-OutR).

% run(Name, State, InL-InR, OutL-OutR)
run(_, final, InL-InR, InL-InR).
run(Name, State, InL-InR, OutL-OutR) :-
    step(Name, InL-InR, State, OutLHalfway-OutRHalfway, OutState),
    run(Name, OutState, OutLHalfway-OutRHalfway, OutL-OutR).
    
% turing(Name, InTape, OutTape)
turing(Name, InTape, OutTape) :-
    run(Name, q0, []-InTape, OutL-OutR),
    reverse_list(OutL, OutLReversed),
    append(OutLReversed, OutR, OutTape).

File: IZPIT/PRETEKLI_IZPITI/2024/2024_3/ts.ml
type state = Q of int | Final
type direction = Left | Right | Stay
type cell = Blank | Zero | One
type tape = { left : cell list; right : cell list }
type program = ((state * cell) * (state * cell * direction)) list

let copy : program =
  [
    ((Q 0, Blank), (Final, Blank, Stay));
    ((Q 0, One), (Q 2, Blank, Right));
    ((Q 2, Blank), (Q 3, Blank, Right));
    ((Q 2, One), (Q 2, One, Right));
    ((Q 3, Blank), (Q 4, One, Left));
    ((Q 3, One), (Q 3, One, Right));
    ((Q 4, Blank), (Q 5, Blank, Left));
    ((Q 4, One), (Q 4, One, Left));
    ((Q 5, Blank), (Q 0, One, Right));
    ((Q 5, One), (Q 5, One, Left));
  ]
let parity : program =
  [
    ((Q 0, Zero), (Q 0, Zero, Right));
    ((Q 0, One), (Q 1, One, Right));
    ((Q 0, Blank), (Final, Zero, Stay));
    ((Q 1, Zero), (Q 1, Zero, Right));
    ((Q 1, One), (Q 0, One, Right));
    ((Q 1, Blank), (Final, One, Stay));
  ]
let plus1 : program =
  [
    ((Q 0, One), (Q 0, One, Right));
    ((Q 0, Blank), (Final, One, Stay));
  ]
let action direction tape : tape =
  let move_left = function
  | { left = []; right } -> { left = []; right = Blank :: right }
  | { left = h :: t; right } -> { left = t; right = h :: right }
  in
  let move_right = function
  | { left; right = [] } -> { left = Blank :: left; right = [] }
  | { left; right = h :: t } -> { left = h :: left; right = t }
  in
  match direction with
  | Stay -> tape
  | Left -> move_left tape
  | Right -> move_right tape



(* type tape = { left : cell list; right : cell list } *)
let read (tape:tape):cell = 
  match tape with
  | {left=_; right=(h::_)} -> h
  | {left=_; right=[]} -> Blank

let write (cell:cell) (tape:tape):tape =
  match tape with
  | {left=_; right=[]} -> {tape with right=[cell]}
  | {left=_; right=(h::t)} -> {tape with right=(cell::t)}

(* type program = ((state * cell) * (state * cell * direction)) list *)
let step (program:program) (state) (tape) =
  match program with
  | [] -> (state, tape)
  | (state, (ns, nc, dir))::_ -> (ns, action dir tape)
